code/controllers/subsystem/xenoarchaeology.dm

```diff
@@ -0,0 +1,186 @@
+	xenoa_majors = compile_artifact_weights(/datum/xenoartifact_trait/major)
+	xenoa_malfunctions = compile_artifact_weights(/datum/xenoartifact_trait/malfunction)
+	xenoa_all_traits = compile_artifact_weights(/datum/xenoartifact_trait)
+	xenoa_all_traits_keyed = compile_artifact_weights(/datum/xenoartifact_trait, TRUE)
+
+	//Traits divided by flavor
+	xenoa_bluespace_traits = compile_artifact_whitelist(/datum/xenoartifact_material/bluespace)
+	xenoa_plasma_traits = compile_artifact_whitelist(/datum/xenoartifact_material/plasma)
+	xenoa_uranium_traits = compile_artifact_whitelist(/datum/xenoartifact_material/uranium)
+	xenoa_bananium_traits = compile_artifact_whitelist(/datum/xenoartifact_material/bananium)
+	xenoa_pearl_traits = compile_artifact_whitelist(/datum/xenoartifact_material/pearl)
+
+	//Compatabilities
+	xenoa_item_incompatible = compile_artifact_compatibilties(TRAIT_INCOMPATIBLE_ITEM)
+	xenoa_mob_incompatible = compile_artifact_compatibilties(TRAIT_INCOMPATIBLE_MOB)
+	xenoa_structure_incompatible = compile_artifact_compatibilties(TRAIT_INCOMPATIBLE_STRUCTURE)
+
+	//Labeler
+	labeler_activator_traits = get_trait_list_stats(xenoa_activators)
+	labeler_minor_traits = get_trait_list_stats(xenoa_minors)
+	labeler_major_traits = get_trait_list_stats(xenoa_majors)
+	labeler_malfunction_traits = get_trait_list_stats(xenoa_malfunctions)
+	labeler_traits_filter = build_trait_filters()
+
+/datum/controller/subsystem/xenoarchaeology/Shutdown()
+	. = ..()
+
+/datum/controller/subsystem/xenoarchaeology/proc/register_console(var/obj/machinery/computer/xenoarchaeology_console/new_console)
+	if(main_console)
+		main_console.main_console = FALSE
+		UnregisterSignal(main_console, COMSIG_PARENT_QDELETING)
+	main_console = new_console
+	main_console.main_console = TRUE
+	RegisterSignal(main_console, COMSIG_PARENT_QDELETING, PROC_REF(catch_console))
+
+/datum/controller/subsystem/xenoarchaeology/proc/catch_console(datum/source)
+	SIGNAL_HANDLER
+
+	main_console = null
+	SEND_SIGNAL(src, XENOA_NEW_CONSOLE)
+
+///Proc used to compile trait weights into a list
+/datum/controller/subsystem/xenoarchaeology/proc/compile_artifact_weights(path, keyed = FALSE)
+	if(!ispath(path))
+		return
+	var/list/temp = subtypesof(path)
+	var/list/weighted = list()
+	for(var/datum/xenoartifact_trait/T as() in temp)
+		if(initial(T.flags) & XENOA_HIDE_TRAIT)
+			continue
+		if(keyed)
+			weighted += list(initial(T.label_name) = (T))
+		else
+			weighted += list((T) = initial(T.rarity)) //The (T) will not work if it is T
+	return weighted
+
+///Compile a blacklist of traits from a given flag/s
+/datum/controller/subsystem/xenoarchaeology/proc/compile_artifact_whitelist(var/flags)
+	var/list/output = list()
+	for(var/datum/xenoartifact_trait/T as() in xenoa_all_traits)
+		if(initial(T.flags) & XENOA_HIDE_TRAIT)
+			continue
+		if(!ispath(flags))
+			if((initial(T.flags) & flags))
+				output += T
+		else
+			var/datum/xenoartifact_material/M = flags
```

Comment:
`var/flags` doesn't sound right for this. Isn't it passing typepath?

Reply:
It's a bit messy, but you can either pass a flag or a material datum, which will load its own flags.